File: Practica 3/Python/pila.py
class Pila:
    def __init__(self):
        self.__elementos = []

    def apilar(self, elemento):
        self.__elementos.append(elemento)

    def desapilar(self):
        if self.esta_vacia():
            raise IndexError("La pila está vacía")
        return self.__elementos.pop()

    def esta_vacia(self):
        return len(self.__elementos) == 0

    def ver_tope(self):
        if self.esta_vacia():
            raise IndexError("La pila está vacía")
        return self.__elementos[-1]

    def __str__(self):
        return str(self.__elementos)


pila = Pila()
pila.apilar(10)
pila.apilar(20)
pila.apilar(30)
print("Pila:", pila)
print("Tope:", pila.ver_tope())
print("Desapiladas:", pila.desapilar())
print("Pilas restantes:", pila)

File: Practica 3/Java/Pila/src/pila/Pila.java
package pila;

public class Pila {
    private int[] elementos;
    private int tope;

    public Pila(int capacidad) {
        elementos = new int[capacidad];
        tope = -1;
    }

    public void apilar(int elemento) {
        if (tope == elementos.length - 1) {
            throw new IllegalStateException("La pila está llena");
        }
        elementos[++tope] = elemento;
    }

    public int desapilar() {
        if (estaVacia()) {
            throw new IllegalStateException("La pila está vacía");
        }
        return elementos[tope--];
    }

    public boolean estaVacia() {
        return tope == -1;
    }

    public int verTope() {
        if (estaVacia()) {
            throw new IllegalStateException("La pila está vacía");
        }
        return elementos[tope];
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i <= tope; i++) {
            sb.append(elementos[i]);
            if (i < tope) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) {
        Pila pila = new Pila(5);
        pila.apilar(10);
        pila.apilar(20);
        pila.apilar(30);
        System.out.println("Pila: " + pila);
        System.out.println("Tope: " + pila.verTope());
        System.out.println("Desapilar: " + pila.desapilar());
        System.out.println("Pila después de desapilar: " + pila);
    }
}
